### PR TITLE
Feature/1156 mq in compound value for subscriptions

### PR DESCRIPTION
**Pull request by kzangeli**
Last PR for #1156 

Subscriptions with filters on metadata compound values

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -38,6 +38,7 @@
 - Fix: use X-Real-IP header as preferred option for from= field in log traces (#2353)
 - Fix: use op= instead of function= in log traces (along with moving the file/line identification from msg= to op=) (#2353)
 - Fix: reorder field in log traces lo align with IoTP Operations requirements (#2353)
+- Fix: bug that provoked old compound values of metadatas to be sent in notifications
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: This problem didn't happen in 1.2.0 (as the whole compound metadata functionality has been recently developed). Thus, this line is not needed.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 2]
```diff
@@ -1824,8 +1824,12 @@ static void updateAttrInNotifyCer
         if (caP->compoundValueP != NULL)
         {
           delete caP->compoundValueP;
+          caP->compoundValueP = NULL;
         }
-        caP->compoundValueP = targetAttr->compoundValueP == NULL ? NULL : targetAttr->compoundValueP->clone();
+
+        // Steal compound value from targetAttr
+        caP->compoundValueP        = targetAttr->compoundValueP;
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: I understand that changing strategy from cloning to stealling may involve some "move" in the release logic. Has you check that?

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 3]
```diff
@@ -1836,21 +1840,36 @@ static void updateAttrInNotifyCer
...
+
+            // Free old value of compound, if any
+            if (mdP->compoundValueP != NULL)
+            {
+              delete mdP->compoundValueP;
+              mdP->compoundValueP = NULL;
+            }
+
+            // Steal compound value from targetMdP
+            mdP->compoundValueP       = targetMdP->compoundValueP;
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Where the compounds are released?

File: test/functionalTest/cases/1156_qfilters_and_compounds/qfilters_and_compounds_deeper.test [hunk 4]
```diff
@@ -653,7 +653,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
                             "F": false, 
                             "N": null, 
                             "T": true, 
-                            "f": 3.14,
+                            "f": 3.14REGEX([01]*),
```
Comment (fgalan) [inline test/functionalTest/cases/1156_qfilters_and_compounds/qfilters_and_compounds_deeper.test]: This is the "rollback" of the .test that I wrongly changed in my previous PR, isn't it?

File: test/functionalTest/cases/2207_not_spurious_decimals_in_custom_notifications/not_spurious_decimals_in_custom_notifications.test [hunk 5]
```diff
@@ -170,15 +170,15 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 The temperature is 34.4 degrees=======================================
```
Comment (fgalan) [inline test/functionalTest/cases/2207_not_spurious_decimals_in_custom_notifications/not_spurious_decimals_in_custom_notifications.test]: Discussed at skype... the changes in this .test should be rolled-back.

File: test/functionalTest/cases/1156_qfilters_and_compounds/qfilters_and_compounds_deeper.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1156_qfilters_and_compounds/qfilters_and_compounds_deeper.test]: Yes

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Yeah, I realized that also in bed last night :-)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Actually, no strategic change is needed, but good that you wondered.
I found out that Metadata::release was empty.
Before implementing compound values for Metadata, there was nothing to delete. Now there is.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Nowhere :-)
In next commit they will be released in Metadata::release()

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in cdc4d46

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in cdc4d46

**Comment by fgalan** [comment]
LGTM

**Pull request merged**